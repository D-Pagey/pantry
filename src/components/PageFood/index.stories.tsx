import React from 'react';

import { UserDan, Fridge, TenantHeidi } from '../../fixtures';
import { AuthContext } from '../ProviderAuth';
import { PageFood } from '.';

export default { title: 'PageFood' };

export const normal = () => (
    <AuthContext.Provider
        value={{
            fetchUserData: () => {},
            isAuthed: true,
            isCheckingAuth: false,
            setUser: () => {},
            signOut: () => {},
            user: UserDan
        }}
    >
        <PageFood fridge={Fridge} tenants={[TenantHeidi]} />
    </AuthContext.Provider>
);
