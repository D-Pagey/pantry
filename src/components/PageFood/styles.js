import styled from 'styled-components';
import { Link } from 'react-router-dom';

import { Button as Btn } from '../Button';

export const Wrapper = styled.div``;

export const RouterLink = styled(Link)`
    color: inherit;
    text-decoration: none;
`;

export const FoodCardWrapper = styled.div`
    align-items: center;
    display: flex;
    flex-direction: column;
`;

export const Button = styled(Btn)`
    justify-self: end;
    margin: 0 1rem;
`;
