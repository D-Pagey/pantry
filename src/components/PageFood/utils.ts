import { DatabaseCategoryType, FoodTypes } from '../../types';

export const filterFridge = (fridge: FoodTypes[], category: DatabaseCategoryType): FoodTypes[] => {
    return fridge.filter((food) => {
        if (food.categories.includes(category.id)) {
            return true;
        }

        return false;
    });
};


// TODO: Refactor this to be one function
export const swapIdsForNames = (fridge: FoodTypes[], categories: DatabaseCategoryType[]): FoodTypes[] => {
    return fridge.map(foodItem => {
        const categoryNames = [...foodItem.categories.map(categoryId => {
            const fullCategory = categories.filter(item => item.id === categoryId);
            return fullCategory[0].name;
        })];
        
        return {
            ...foodItem,
            categories: categoryNames
        };
    });
};

export const swapNamesForIds = (fridge: FoodTypes[], categories: DatabaseCategoryType[]): FoodTypes[] => {
    return fridge.map(foodItem => {
        const categoryIds = [...foodItem.categories.map(categoryName => {
            const fullCategory = categories.filter(item => item.name === categoryName);
            return fullCategory[0].id;
        })];
        
        return {
            ...foodItem,
            categories: categoryIds
        };
    });
};