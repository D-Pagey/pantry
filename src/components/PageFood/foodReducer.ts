import { FoodType, TenantType } from '../../types';
import { getExpiringItems } from '../../utils';
import { applyMultipleFilters } from './utils';

export type SortOptions = 'name' | 'date';

export type FilterState = {
    category: string;
    selectedOwners: string[];
    showOnlyExpiring: boolean;
    sortBy: SortOptions;
};

type ApplyFiltersAction = {
    type: 'APPLY_FILTERS';
    fridge: FoodType[];
};

type RemoveExpiringFilterAction = {
    type: 'REMOVE_EXPIRING_FILTER';
    fridge: FoodType[];
};

type ChangeSortAction = {
    type: 'CHANGE_SORTED_BY';
    sortBy: SortOptions;
};

type RemoveSelectedOwnersAction = {
    type: 'REMOVE_SELECTED_OWNERS';
    fridge: FoodType[];
    tenants: TenantType[];
};

type UpdateFridgeAction = {
    type: 'UPDATE_FRIDGE';
    fridge: FoodType[];
};

type RemoveCategoryAction = {
    type: 'REMOVE_CATEGORY';
    fridge: FoodType[];
};

type ToggleOwnerAction = {
    type: 'TOGGLE_SELECTED_OWNER';
    ownerId: string;
};

type ChangeCategoryAction = {
    type: 'CHANGE_CATEGORY';
    category: string;
};

type ResetFiltersAction = {
    type: 'RESET';
};

type ChangeShowExpiredAction = {
    type: 'CHANGE_SHOW_ONLY_EXPIRED';
    onlyShowExpired: boolean;
};

export type FoodActions =
    | ApplyFiltersAction
    | RemoveExpiringFilterAction
    | UpdateFridgeAction
    | RemoveSelectedOwnersAction
    | RemoveCategoryAction
    | ToggleOwnerAction
    | ChangeCategoryAction
    | ResetFiltersAction
    | ChangeShowExpiredAction
    | ChangeSortAction;

export type FoodState = {
    food: FoodType[];
    appliedFilters: FilterState;
    pendingFilters: FilterState;
};

export const init = (initialFoodState: FoodState, tenants: TenantType[], fridge: FoodType[]): FoodState => {
    const expiringItems = getExpiringItems(fridge);

    const filters = {
        ...initialFoodState.appliedFilters,
        showOnlyExpiring: expiringItems.length > 0,
        selectedOwners: tenants.map((tenant) => tenant.uid)
    };

    return {
        ...initialFoodState,
        appliedFilters: filters,
        pendingFilters: filters,
        food: applyMultipleFilters(fridge, filters)
    };
};

const defaultFilters: FilterState = {
    selectedOwners: [],
    showOnlyExpiring: true,
    sortBy: 'date',
    category: ''
};

export const initialFoodState: FoodState = {
    food: [],
    appliedFilters: defaultFilters,
    pendingFilters: defaultFilters
};

export const foodReducer = (state: FoodState, action: FoodActions): FoodState => {
    switch (action.type) {
        case 'APPLY_FILTERS': {
            return {
                ...state,
                appliedFilters: state.pendingFilters,
                food: applyMultipleFilters(action.fridge, state.pendingFilters)
            };
        }

        case 'REMOVE_EXPIRING_FILTER': {
            const updatedFilters = {
                ...state.appliedFilters,
                showOnlyExpiring: false
            };

            return {
                ...state,
                appliedFilters: updatedFilters,
                pendingFilters: updatedFilters,
                food: applyMultipleFilters(action.fridge, updatedFilters)
            };
        }

        case 'REMOVE_SELECTED_OWNERS': {
            const updatedFilters = {
                ...state.appliedFilters,
                selectedOwners: action.tenants.map((tenant) => tenant.uid)
            };

            return {
                ...state,
                appliedFilters: updatedFilters,
                pendingFilters: updatedFilters,
                food: applyMultipleFilters(action.fridge, updatedFilters)
            };
        }

        case 'REMOVE_CATEGORY': {
            const updatedFilters = {
                ...state.appliedFilters,
                category: ''
            };

            return {
                ...state,
                appliedFilters: updatedFilters,
                pendingFilters: updatedFilters,
                food: applyMultipleFilters(action.fridge, updatedFilters)
            };
        }

        case 'UPDATE_FRIDGE': {
            return {
                ...state,
                food: applyMultipleFilters(action.fridge, state.appliedFilters)
            };
        }

        case 'TOGGLE_SELECTED_OWNER': {
            const {
                pendingFilters: { selectedOwners }
            } = state;

            const updatedSelectedOwners = selectedOwners.includes(action.ownerId)
                ? selectedOwners.filter((id) => id !== action.ownerId)
                : [...selectedOwners, action.ownerId];

            return {
                ...state,
                pendingFilters: {
                    ...state.pendingFilters,
                    selectedOwners: updatedSelectedOwners
                }
            };
        }

        case 'CHANGE_CATEGORY': {
            return {
                ...state,
                pendingFilters: {
                    ...state.pendingFilters,
                    category: action.category
                }
            };
        }

        case 'RESET': {
            return {
                ...state,
                pendingFilters: state.appliedFilters
            };
        }

        case 'CHANGE_SHOW_ONLY_EXPIRED': {
            return {
                ...state,
                pendingFilters: {
                    ...state.pendingFilters,
                    showOnlyExpiring: action.onlyShowExpired
                }
            };
        }

        case 'CHANGE_SORTED_BY': {
            return {
                ...state,
                pendingFilters: {
                    ...state.pendingFilters,
                    sortBy: action.sortBy
                }
            };
        }

        default:
            return state;
    }
};
