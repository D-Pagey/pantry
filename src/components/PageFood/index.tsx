import React, { FC, useCallback, useContext, useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useMediaQuery } from 'react-responsive';
import arraySort from 'array-sort';
import { toast } from 'react-toastify';

import { db } from '../../services';
import { FoodType, TenantType } from '../../types';
import { getExpiringItems, filterFridgeByCategory, getCategoriesAndCounts, filterByTenantIds } from '../../utils';
import { mediaQuery } from '../../tokens';
import { Layout } from '../Layout';
import { CategoryFilterMobile } from '../CategoryFilterMobile';
import { CategoryFilterDesktop } from '../CategoryFilterDesktop';
import { OwnerFilter } from '../OwnerFilter';
import { FoodCard } from '../FoodCard';
import { FoodOptions } from '../FoodOptions';
import { AuthContext } from '../ProviderAuth';
import * as S from './styles';

type PageFoodProps = {
    fridge: FoodType[];
    tenants: TenantType[];
};

export const PageFood: FC<PageFoodProps> = ({ fridge, tenants }) => {
    const [selectedFood, setSelectedFood] = useState<FoodType[]>();
    const [category, setCategory] = useState('all');
    const [isExpiring, setIsExpiring] = useState(false);
    const [editingItem, setEditingItem] = useState<FoodType | undefined>();
    const [selectedTenants, setSelectedTenants] = useState<string[]>([]);
    const { user } = useContext(AuthContext);
    const history = useHistory();
    const isTabletOrLarger = useMediaQuery({
        query: mediaQuery.tablet
    });

    const deleteFoodItem = (name: string): void => {
        if (user) {
            db.collection('households')
                .doc(user.household)
                .update({ [`fridge.${name}.batches`]: {} })
                .then(() => toast.error('Food deleted'))
                .catch(() => toast.error('Error with deleting food'));
        }
    };

    const filterFood = useCallback(() => {
        const isCategoryAll = category === 'all';

        let filteredFridge = [...fridge];

        if (!isCategoryAll) {
            filteredFridge = filterFridgeByCategory(filteredFridge, category);
        }

        if (isExpiring) {
            filteredFridge = getExpiringItems(filteredFridge);
        }

        if (selectedTenants.length > 0) {
            filteredFridge = filterByTenantIds(filteredFridge, selectedTenants);
        }

        return filteredFridge;
    }, [category, fridge, isExpiring, selectedTenants]);

    useEffect(() => {
        if (fridge.length > 0 && tenants.length > 0) {
            setSelectedFood(filterFood());
        }
    }, [category, selectedTenants, isExpiring, filterFood, fridge, tenants]);

    const handleCategoryClick = (selectedCategory: string): void => {
        setCategory(selectedCategory);
    };

    const handleExpiringClick = (): void => {
        setIsExpiring(!isExpiring);
    };

    const handleFoodDelete = (): void => {
        if (editingItem?.name) {
            deleteFoodItem(editingItem.name);
            setEditingItem(undefined);
        }
    };

    const handleFoodEdit = (): void => {
        history.push(`/${editingItem?.name}/edit`);
    };

    const handleFoodClick = (item: FoodType) => (): void => {
        if (!editingItem || editingItem.name !== item.name) setEditingItem(item);
        if (editingItem?.name === item.name) setEditingItem(undefined);
    };

    const handleAddClick = () => history.push('/add');

    return (
        <Layout title="Your Food:">
            <S.Wrapper>
                {isTabletOrLarger ? (
                    <CategoryFilterDesktop
                        categories={getCategoriesAndCounts(fridge)}
                        selected={category}
                        handleCategoryClick={handleCategoryClick}
                    />
                ) : (
                    <CategoryFilterMobile selected={category} setSelected={handleCategoryClick} />
                )}

                <S.FilterWrapper>
                    <OwnerFilter
                        tenants={tenants}
                        setSelectedTenants={setSelectedTenants}
                        selectedTenants={selectedTenants}
                    />

                    <S.ExpiringButton handleClick={handleExpiringClick} isEnabled={isExpiring} />
                </S.FilterWrapper>

                {fridge?.length === 0 && <p data-testid="pageFoodNoData">You have no food in your fridge.</p>}

                {fridge?.length !== 0 && selectedFood?.length === 0 && (
                    <p data-testid={`pageFoodNoData${category}`}>
                        There is no {isExpiring && 'expiring'} food that falls under the category of {category}
                    </p>
                )}

                <S.FoodCardGrid>
                    {selectedFood &&
                        arraySort(selectedFood, 'name').map((item: FoodType) => {
                            if (item.batches.length > 0) {
                                return (
                                    <FoodCard
                                        handleClick={handleFoodClick(item)}
                                        isSelected={item.name === editingItem?.name}
                                        item={item}
                                        key={item.name}
                                        margin="0 0 1rem"
                                        tenants={tenants}
                                    />
                                );
                            }

                            return null;
                        })}
                </S.FoodCardGrid>

                <S.AddButton size="sm" onClick={handleAddClick}>
                    Add Item
                </S.AddButton>
            </S.Wrapper>

            {editingItem && (
                <FoodOptions name={editingItem.name} handleDelete={handleFoodDelete} handleEdit={handleFoodEdit} />
            )}
        </Layout>
    );
};
