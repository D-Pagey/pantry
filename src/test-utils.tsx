import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
