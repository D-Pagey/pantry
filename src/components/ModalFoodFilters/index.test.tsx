import { render } from '../../test-utils';
import { Categories, TenantDan, TenantJoe } from '../../fixtures';
import { SortOptions } from '../PageFood/foodReducer';
import { ModalFoodFilters } from '.';

const props = {
    categories: Categories,
    dispatch: () => null,
    filters: {
        category: '',
        sortBy: 'name' as SortOptions,
        showOnlyExpiring: false,
        selectedOwners: [TenantDan.uid, TenantJoe.uid]
    },
    handleApplyFiltersClick: () => null,
    handleCancelClick: () => null,
    isModalOpen: false,
    tenants: [TenantDan, TenantJoe]
};

describe('ModalFoodFilters component', () => {
    it('should render', () => {
        const { container } = render(<ModalFoodFilters {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
