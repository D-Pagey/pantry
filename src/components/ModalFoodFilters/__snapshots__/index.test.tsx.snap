// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalFoodFilters component should render 1`] = `null`;
